File: go/fragment.go
package topayz512

import (
	"encoding/binary"
	"sync"
	"time"
)

// Data fragmentation for TOPAY-Z512 with optimizations

// Fragment represents a single data fragment
type Fragment struct {
	ID       uint32 `json:"id"`
	Index    uint32 `json:"index"`
	Total    uint32 `json:"total"`
	Data     []byte `json:"data"`
	Checksum Hash   `json:"checksum"`
	Size     uint32 `json:"size"`
}

// FragmentationResult contains the result of data fragmentation
type FragmentationResult struct {
	Fragments    []Fragment       `json:"fragments"`
	TotalSize    uint64           `json:"total_size"`
	FragmentSize uint32           `json:"fragment_size"`
	Metadata     FragmentMetadata `json:"metadata"`
}

// FragmentMetadata contains metadata about the fragmentation
type FragmentMetadata struct {
	OriginalSize  uint64    `json:"original_size"`
	FragmentCount uint32    `json:"fragment_count"`
	Timestamp     time.Time `json:"timestamp"`
	Algorithm     string    `json:"algorithm"`
	Checksum      Hash      `json:"checksum"`
}

// ReconstructionResult contains the result of data reconstruction
type ReconstructionResult struct {
	Data         []byte           `json:"data"`
	IsComplete   bool             `json:"is_complete"`
	MissingCount uint32           `json:"missing_count"`
	Metadata     FragmentMetadata `json:"metadata"`
}

// ShouldFragment determines if data should be fragmented based on size
func ShouldFragment(dataSize int) bool {
	return dataSize >= MinFragmentThreshold
}

// CalculateFragmentCount calculates the optimal number of fragments
func CalculateFragmentCount(dataSize int) int {
	if !ShouldFragment(dataSize) {
		return 1
	}

	fragmentCount := (dataSize + FragmentSize - 1) / FragmentSize
	if fragmentCount > MaxFragments {
		fragmentCount = MaxFragments
	}

	return fragmentCount
}

// FragmentData splits data into fragments for parallel processing
func FragmentData(data []byte) (FragmentationResult, error) {
	if len(data) == 0 {
		return FragmentationResult{}, ErrEmptyData
	}

	fragmentCount := CalculateFragmentCount(len(data))
	fragmentSize := (len(data) + fragmentCount - 1) / fragmentCount

	// Generate unique fragment ID
	idBytes, err := SecureRandom(4)
	if err != nil {
		return FragmentationResult{}, err
	}
	fragmentID := binary.BigEndian.Uint32(idBytes)

	// Calculate total checksum
	totalChecksum := ComputeHash(data)

	// Create fragments
	fragments := make([]Fragment, fragmentCount)

	for i := 0; i < fragmentCount; i++ {
		start := i * fragmentSize
		end := start + fragmentSize
		if end > len(data) {
			end = len(data)
		}

		fragmentData := make([]byte, end-start)
		copy(fragmentData, data[start:end])

		// Calculate fragment checksum
		fragmentChecksum := ComputeHash(fragmentData)

		fragments[i] = Fragment{
			ID:       fragmentID,
			Index:    uint32(i),
			Total:    uint32(fragmentCount),
			Data:     fragmentData,
			Checksum: fragmentChecksum,
		}
	}

	metadata := FragmentMetadata{
		OriginalSize:  uint64(len(data)),
		FragmentCount: uint32(fragmentCount),
		Timestamp:     time.Now(),
		Algorithm:     "TOPAY-Z512",
		Checksum:      totalChecksum,
	}

	return FragmentationResult{
		Fragments: fragments,
		Metadata:  metadata,
	}, nil
}

// ReconstructData reconstructs original data from fragments
func ReconstructData(fragments []Fragment) (ReconstructionResult, error) {
	if len(fragments) == 0 {
		return ReconstructionResult{}, ErrEmptyData
	}

	// Validate fragments
	fragmentID := fragments[0].ID
	totalFragments := fragments[0].Total

	if len(fragments) != int(totalFragments) {
		return ReconstructionResult{}, ErrInvalidFragmentCount
	}

	// Sort fragments by index
	sortedFragments := make([]Fragment, len(fragments))
	copy(sortedFragments, fragments)

	// Simple bubble sort for fragment ordering
	for i := 0; i < len(sortedFragments); i++ {
		for j := 0; j < len(sortedFragments)-1-i; j++ {
			if sortedFragments[j].Index > sortedFragments[j+1].Index {
				sortedFragments[j], sortedFragments[j+1] = sortedFragments[j+1], sortedFragments[j]
			}
		}
	}

	// Validate fragment integrity
	for i, fragment := range sortedFragments {
		if fragment.ID != fragmentID {
			return ReconstructionResult{}, ErrReconstructionFailed
		}

		if fragment.Index != uint32(i) {
			return ReconstructionResult{}, ErrReconstructionFailed
		}

		if fragment.Total != totalFragments {
			return ReconstructionResult{}, ErrReconstructionFailed
		}

		// Verify fragment checksum
		computedChecksum := ComputeHash(fragment.Data)
		if !HashEqual(computedChecksum, fragment.Checksum) {
			return ReconstructionResult{}, ErrReconstructionFailed
		}
	}

	// Reconstruct data
	var totalSize int
	for _, fragment := range sortedFragments {
		totalSize += len(fragment.Data)
	}

	reconstructedData := make([]byte, 0, totalSize)
	for _, fragment := range sortedFragments {
		reconstructedData = append(reconstructedData, fragment.Data...)
	}

	// Verify total checksum
	totalChecksum := ComputeHash(reconstructedData)

	metadata := FragmentMetadata{
		OriginalSize:  uint64(len(reconstructedData)),
		FragmentCount: totalFragments,
		Timestamp:     time.Now(),
		Algorithm:     "TOPAY-Z512",
		Checksum:      totalChecksum,
	}

	return ReconstructionResult{
		Data:         reconstructedData,
		IsComplete:   true,
		MissingCount: 0,
		Metadata:     metadata,
	}, nil
}

// Parallel fragmentation operations

// ParallelFragmentData fragments data using parallel processing
func ParallelFragmentData(data []byte) (FragmentationResult, error) {
	if len(data) == 0 {
		return FragmentationResult{}, ErrEmptyData
	}

	fragmentCount := CalculateFragmentCount(len(data))
	fragmentSize := (len(data) + fragmentCount - 1) / fragmentCount

	// Generate unique fragment ID
	idBytes, err := SecureRandom(4)
	if err != nil {
		return FragmentationResult{}, err
	}
	fragmentID := binary.BigEndian.Uint32(idBytes)

	// Calculate total checksum
	totalChecksum := ComputeHash(data)

	// Create fragments in parallel
	fragments := make([]Fragment, fragmentCount)

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > fragmentCount {
		numWorkers = fragmentCount
	}

	// Channel for work distribution
	workChan := make(chan int, fragmentCount)
	resultChan := make(chan struct {
		index    int
		fragment Fragment
		err      error
	}, fragmentCount)

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				start := index * fragmentSize
				end := start + fragmentSize
				if end > len(data) {
					end = len(data)
				}

				fragmentData := make([]byte, end-start)
				copy(fragmentData, data[start:end])

				// Calculate fragment checksum
				fragmentChecksum := ComputeHash(fragmentData)

				fragment := Fragment{
					ID:       fragmentID,
					Index:    uint32(index),
					Total:    uint32(fragmentCount),
					Data:     fragmentData,
					Checksum: fragmentChecksum,
				}

				resultChan <- struct {
					index    int
					fragment Fragment
					err      error
				}{index, fragment, nil}
			}
		}()
	}

	// Send work
	go func() {
		for i := 0; i < fragmentCount; i++ {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.err != nil {
			return FragmentationResult{}, result.err
		}
		fragments[result.index] = result.fragment
	}

	metadata := FragmentMetadata{
		OriginalSize:  uint64(len(data)),
		FragmentCount: uint32(fragmentCount),
		Timestamp:     time.Now(),
		Algorithm:     "TOPAY-Z512",
		Checksum:      totalChecksum,
	}

	return FragmentationResult{
		Fragments: fragments,
		Metadata:  metadata,
	}, nil
}

// ParallelReconstructData reconstructs data using parallel processing
func ParallelReconstructData(fragments []Fragment) (ReconstructionResult, error) {
	if len(fragments) == 0 {
		return ReconstructionResult{}, ErrEmptyData
	}

	// Validate and sort fragments first
	result, err := ReconstructData(fragments)
	if err != nil {
		return result, err
	}

	// For parallel reconstruction, we could verify checksums in parallel
	// but since reconstruction is already fast, we'll use the sequential version
	return result, nil
}

// Fragment serialization

// SerializeFragment converts a fragment to bytes
func SerializeFragment(fragment Fragment) []byte {
	// Calculate total size needed
	dataLen := len(fragment.Data)
	totalSize := 4 + 4 + 4 + 4 + dataLen + HashSize // ID + Index + Total + DataLen + Data + Checksum

	result := make([]byte, totalSize)
	offset := 0

	// Write ID
	binary.BigEndian.PutUint32(result[offset:], fragment.ID)
	offset += 4

	// Write Index
	binary.BigEndian.PutUint32(result[offset:], fragment.Index)
	offset += 4

	// Write Total
	binary.BigEndian.PutUint32(result[offset:], fragment.Total)
	offset += 4

	// Write Data length
	binary.BigEndian.PutUint32(result[offset:], uint32(dataLen))
	offset += 4

	// Write Data
	copy(result[offset:], fragment.Data)
	offset += dataLen

	// Write Checksum
	copy(result[offset:], fragment.Checksum[:])

	return result
}

// DeserializeFragment converts bytes to a fragment
func DeserializeFragment(data []byte) (Fragment, error) {
	if len(data) < 16+HashSize { // Minimum size: 4 fields + checksum
		return Fragment{}, ErrInvalidFragmentCount
	}

	offset := 0

	// Read ID
	id := binary.BigEndian.Uint32(data[offset:])
	offset += 4

	// Read Index
	index := binary.BigEndian.Uint32(data[offset:])
	offset += 4

	// Read Total
	total := binary.BigEndian.Uint32(data[offset:])
	offset += 4

	// Read Data length
	dataLen := binary.BigEndian.Uint32(data[offset:])
	offset += 4

	if len(data) < offset+int(dataLen)+HashSize {
		return Fragment{}, ErrInvalidFragmentCount
	}

	// Read Data
	fragmentData := make([]byte, dataLen)
	copy(fragmentData, data[offset:offset+int(dataLen)])
	offset += int(dataLen)

	// Read Checksum
	var checksum Hash
	copy(checksum[:], data[offset:])

	return Fragment{
		ID:       id,
		Index:    index,
		Total:    total,
		Data:     fragmentData,
		Checksum: checksum,
	}, nil
}

// Mobile optimization

// MobileLatencyEstimate estimates processing latency for mobile devices
type MobileLatencyEstimate struct {
	FragmentationMs   float64
	ReconstructionMs  float64
	TotalMs           float64
	RecommendedChunks int
}

// EstimateMobileLatency estimates processing time for mobile devices
func EstimateMobileLatency(dataSize int) MobileLatencyEstimate {
	// Base latency factors for mobile devices (conservative estimates)
	const (
		baseFragmentationMsPerKB  = 0.1
		baseReconstructionMsPerKB = 0.05
		mobileCPUFactor           = 2.0 // Mobile CPUs are typically 2x slower
	)

	dataSizeKB := float64(dataSize) / 1024.0

	fragmentationMs := dataSizeKB * baseFragmentationMsPerKB * mobileCPUFactor
	reconstructionMs := dataSizeKB * baseReconstructionMsPerKB * mobileCPUFactor
	totalMs := fragmentationMs + reconstructionMs

	// Recommend optimal chunk count for mobile
	recommendedChunks := CalculateFragmentCount(dataSize)
	if recommendedChunks > 64 { // Limit for mobile devices
		recommendedChunks = 64
	}

	return MobileLatencyEstimate{
		FragmentationMs:   fragmentationMs,
		ReconstructionMs:  reconstructionMs,
		TotalMs:           totalMs,
		RecommendedChunks: recommendedChunks,
	}
}

// Fragmented cryptographic operations

// FragmentedHash computes hash using fragmented processing
func FragmentedHash(data []byte) (Hash, error) {
	if !ShouldFragment(len(data)) {
		return ComputeHash(data), nil
	}

	fragResult, err := ParallelFragmentData(data)
	if err != nil {
		return Hash{}, err
	}

	// Compute hashes of fragments in parallel
	fragmentHashes := make([]Hash, len(fragResult.Fragments))

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > len(fragResult.Fragments) {
		numWorkers = len(fragResult.Fragments)
	}

	// Channel for work distribution
	workChan := make(chan int, len(fragResult.Fragments))
	resultChan := make(chan struct {
		index int
		hash  Hash
	}, len(fragResult.Fragments))

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				hash := ComputeHash(fragResult.Fragments[index].Data)
				resultChan <- struct {
					index int
					hash  Hash
				}{index, hash}
			}
		}()
	}

	// Send work
	go func() {
		for i := range fragResult.Fragments {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		fragmentHashes[result.index] = result.hash
	}

	// Combine fragment hashes
	hs := NewHashState()
	for _, fragmentHash := range fragmentHashes {
		hs.Update(fragmentHash[:])
	}

	return hs.Finalize(), nil
}

// FragmentedKEM performs KEM operations on fragmented data
func FragmentedKEM(data []byte) ([]Ciphertext, []SharedSecret, error) {
	if !ShouldFragment(len(data)) {
		// For small data, use single KEM operation
		publicKey, _, err := KEMKeyGen()
		if err != nil {
			return nil, nil, err
		}

		ciphertext, sharedSecret, err := KEMEncapsulate(publicKey)
		if err != nil {
			return nil, nil, err
		}

		return []Ciphertext{ciphertext}, []SharedSecret{sharedSecret}, nil
	}

	fragResult, err := ParallelFragmentData(data)
	if err != nil {
		return nil, nil, err
	}

	// Generate KEM key pairs for each fragment
	publicKeys, _, err := BatchKEMKeyGen(len(fragResult.Fragments))
	if err != nil {
		return nil, nil, err
	}

	// Perform batch encapsulation
	ciphertexts, sharedSecrets, err := BatchKEMEncapsulate(publicKeys)
	if err != nil {
		return nil, nil, err
	}

	return ciphertexts, sharedSecrets, nil
}

// Performance benchmarking

// FragmentationBenchmark represents fragmentation performance metrics
type FragmentationBenchmark struct {
	ThroughputMBps   float64
	FragmentationMs  float64
	ReconstructionMs float64
	ParallelSpeedup  float64
	MobileLatencyMs  float64
}

// BenchmarkFragmentation measures fragmentation performance
func BenchmarkFragmentation(dataSize int) FragmentationBenchmark {
	// Generate test data
	data := make([]byte, dataSize)
	for i := range data {
		data[i] = byte(i)
	}

	// Sequential fragmentation benchmark
	start := time.Now()
	fragResult, _ := FragmentData(data)
	seqFragDuration := time.Since(start)

	// Sequential reconstruction benchmark
	start = time.Now()
	_, _ = ReconstructData(fragResult.Fragments)
	seqReconDuration := time.Since(start)

	// Parallel fragmentation benchmark
	start = time.Now()
	_, _ = ParallelFragmentData(data)
	parFragDuration := time.Since(start)

	// Calculate metrics
	totalBytes := float64(dataSize)
	throughputMBps := totalBytes / seqFragDuration.Seconds() / (1024 * 1024)
	fragmentationMs := seqFragDuration.Seconds() * 1000
	reconstructionMs := seqReconDuration.Seconds() * 1000
	parallelSpeedup := seqFragDuration.Seconds() / parFragDuration.Seconds()

	// Mobile latency estimate
	mobileEst := EstimateMobileLatency(dataSize)

	return FragmentationBenchmark{
		ThroughputMBps:   throughputMBps,
		FragmentationMs:  fragmentationMs,
		ReconstructionMs: reconstructionMs,
		ParallelSpeedup:  parallelSpeedup,
		MobileLatencyMs:  mobileEst.TotalMs,
	}
}

// Fragment integrity and validation

// ValidateFragmentIntegrity validates the integrity of a fragment
func ValidateFragmentIntegrity(fragment Fragment) error {
	// Verify checksum
	computedChecksum := ComputeHash(fragment.Data)
	if !HashEqual(computedChecksum, fragment.Checksum) {
		return ErrReconstructionFailed
	}

	// Validate fragment parameters
	if fragment.Index >= fragment.Total {
		return ErrInvalidFragmentCount
	}

	if len(fragment.Data) == 0 {
		return ErrEmptyData
	}

	return nil
}

// RepairFragment attempts to repair a corrupted fragment
func RepairFragment(fragment Fragment, originalData []byte, fragmentSize int) (Fragment, error) {
	// Calculate expected fragment data
	start := int(fragment.Index) * fragmentSize
	end := start + fragmentSize
	if end > len(originalData) {
		end = len(originalData)
	}

	if start >= len(originalData) {
		return Fragment{}, ErrInvalidFragmentCount
	}

	// Create repaired fragment
	repairedData := make([]byte, end-start)
	copy(repairedData, originalData[start:end])

	repairedChecksum := ComputeHash(repairedData)

	repairedFragment := Fragment{
		ID:       fragment.ID,
		Index:    fragment.Index,
		Total:    fragment.Total,
		Data:     repairedData,
		Checksum: repairedChecksum,
	}

	return repairedFragment, nil
}

File: js/src/fragment.ts
/**
 * Data fragmentation and reconstruction for TOPAY-Z512
 */

import {
  FRAGMENT_SIZE,
  MIN_FRAGMENT_THRESHOLD,
  MAX_FRAGMENTS,
  Fragment,
  FragmentMetadata,
  FragmentationResult,
  ReconstructionResult,
  EmptyDataError,
  FragmentationFailedError,
  ReconstructionFailedError
  // InvalidFragmentCountError
} from './index';
import { computeHash } from './hash';
import { timestamp, /* copyBytes, */ constantTimeEqual } from './utils';

/**
 * Fragments data into smaller chunks for parallel processing
 * @param data - Data to fragment
 * @param fragmentSize - Size of each fragment (default: FRAGMENT_SIZE)
 * @returns Promise resolving to fragmentation result
 * @throws EmptyDataError if data is empty
 * @throws FragmentationFailedError if fragmentation fails
 */
export async function fragmentData(
  data: Uint8Array,
  fragmentSize: number = FRAGMENT_SIZE
): Promise<FragmentationResult> {
  if (data.length === 0) {
    throw new EmptyDataError();
  }

  if (fragmentSize <= 0) {
    throw new Error('Fragment size must be positive');
  }

  // Calculate number of fragments needed
  const fragmentCount = Math.ceil(data.length / fragmentSize);

  if (fragmentCount > MAX_FRAGMENTS) {
    throw new FragmentationFailedError();
  }

  // Create metadata
  const checksum = await computeHash(data);
  const metadata: FragmentMetadata = {
    originalSize: data.length,
    fragmentCount,
    timestamp: timestamp(),
    algorithm: 'TOPAY-Z512-FRAG-v1',
    checksum
  };

  // Create fragments
  const fragments: Fragment[] = [];

  for (let i = 0; i < fragmentCount; i++) {
    const start = i * fragmentSize;
    const end = Math.min(start + fragmentSize, data.length);
    const fragmentData = data.slice(start, end);

    fragments.push({
      index: i,
      data: fragmentData,
      metadata
    });
  }

  return {
    fragments,
    metadata
  };
}

/**
 * Reconstructs data from fragments
 * @param fragments - Array of fragments to reconstruct
 * @returns Promise resolving to reconstruction result
 * @throws ReconstructionFailedError if reconstruction fails
 */
export async function reconstructData(fragments: Fragment[]): Promise<ReconstructionResult> {
  if (fragments.length === 0) {
    throw new ReconstructionFailedError();
  }

  // Get metadata from first fragment
  const metadata = fragments[0]!.metadata;
  const expectedCount = metadata.fragmentCount;

  // Check if we have all fragments
  const receivedIndices = new Set(fragments.map(f => f.index));
  const missingCount = expectedCount - receivedIndices.size;
  const isComplete = missingCount === 0;

  if (!isComplete) {
    return {
      data: new Uint8Array(0),
      isComplete: false,
      missingCount,
      metadata
    };
  }

  // Sort fragments by index
  const sortedFragments = [...fragments].sort((a, b) => a.index - b.index);

  // Validate fragment sequence
  for (let i = 0; i < sortedFragments.length; i++) {
    if (sortedFragments[i]!.index !== i) {
      throw new ReconstructionFailedError();
    }
  }

  // Reconstruct data
  const reconstructedData = new Uint8Array(metadata.originalSize);
  let offset = 0;

  for (const fragment of sortedFragments) {
    const fragmentData = fragment.data;
    const copyLength = Math.min(fragmentData.length, metadata.originalSize - offset);

    reconstructedData.set(fragmentData.slice(0, copyLength), offset);
    offset += copyLength;
  }

  // Verify checksum
  const computedChecksum = await computeHash(reconstructedData);
  if (!constantTimeEqual(computedChecksum, metadata.checksum)) {
    throw new ReconstructionFailedError();
  }

  return {
    data: reconstructedData,
    isComplete: true,
    missingCount: 0,
    metadata
  };
}

/**
 * Performs parallel fragmentation of multiple data items
 * @param dataItems - Array of data to fragment
 * @param fragmentSize - Size of each fragment
 * @returns Promise resolving to array of fragmentation results
 */
export async function parallelFragmentation(
  dataItems: Uint8Array[],
  fragmentSize: number = FRAGMENT_SIZE
): Promise<FragmentationResult[]> {
  const promises = dataItems.map(data => fragmentData(data, fragmentSize));
  return Promise.all(promises);
}

/**
 * Performs parallel reconstruction of multiple fragment sets
 * @param fragmentSets - Array of fragment arrays to reconstruct
 * @returns Promise resolving to array of reconstruction results
 */
export async function parallelReconstruction(
  fragmentSets: Fragment[][]
): Promise<ReconstructionResult[]> {
  const promises = fragmentSets.map(fragments => reconstructData(fragments));
  return Promise.all(promises);
}

/**
 * Validates fragment integrity
 * @param fragment - Fragment to validate
 * @returns Promise resolving to true if fragment is valid
 */
export async function validateFragment(fragment: Fragment): Promise<boolean> {
  try {
    // Check basic structure
    if (!fragment.data || !fragment.metadata) {
      return false;
    }

    // Check index bounds
    if (fragment.index < 0 || fragment.index >= fragment.metadata.fragmentCount) {
      return false;
    }

    // Check algorithm
    if (fragment.metadata.algorithm !== 'TOPAY-Z512-FRAG-v1') {
      return false;
    }

    return true;
  } catch {
    return false;
  }
}

/**
 * Estimates mobile device latency for fragment processing
 * @param dataSize - Size of data to process
 * @param fragmentSize - Size of each fragment
 * @returns Estimated latency in milliseconds
 */
export function estimateMobileLatency(
  dataSize: number,
  fragmentSize: number = FRAGMENT_SIZE
): number {
  const fragmentCount = Math.ceil(dataSize / fragmentSize);

  // Base latency estimates for mobile devices (in ms)
  const baseFragmentationLatency = 2; // Per fragment
  const baseReconstructionLatency = 1.5; // Per fragment
  const networkLatency = 50; // Network overhead

  const fragmentationTime = fragmentCount * baseFragmentationLatency;
  const reconstructionTime = fragmentCount * baseReconstructionLatency;

  return fragmentationTime + reconstructionTime + networkLatency;
}

/**
 * Determines optimal fragment size for given data and constraints
 * @param dataSize - Size of data to fragment
 * @param maxLatency - Maximum acceptable latency in milliseconds
 * @param deviceType - Type of device ('mobile', 'desktop', 'iot')
 * @returns Optimal fragment size
 */
export function getOptimalFragmentSize(
  dataSize: number,
  maxLatency: number = 100,
  deviceType: 'mobile' | 'desktop' | 'iot' = 'mobile'
): number {
  // Device-specific constraints
  const constraints = {
    mobile: { minSize: 128, maxSize: 512, processingPower: 1.0 },
    desktop: { minSize: 256, maxSize: 2048, processingPower: 3.0 },
    iot: { minSize: 64, maxSize: 256, processingPower: 0.5 }
  };

  const constraint = constraints[deviceType];

  // Start with default fragment size
  let fragmentSize = FRAGMENT_SIZE;

  // Adjust based on data size
  if (dataSize < MIN_FRAGMENT_THRESHOLD) {
    return Math.min(dataSize, constraint.maxSize);
  }

  // Adjust based on latency requirements
  let estimatedLatency = estimateMobileLatency(dataSize, fragmentSize);

  while (estimatedLatency > maxLatency && fragmentSize > constraint.minSize) {
    fragmentSize = Math.max(fragmentSize * 0.8, constraint.minSize);
    estimatedLatency = estimateMobileLatency(dataSize, fragmentSize);
  }

  // Ensure within device constraints
  fragmentSize = Math.max(constraint.minSize, Math.min(fragmentSize, constraint.maxSize));

  return Math.floor(fragmentSize);
}

/**
 * Serializes fragments to a portable format
 * @param fragments - Fragments to serialize
 * @returns Serialized fragments as JSON string
 */
export function serializeFragments(fragments: Fragment[]): string {
  const serializable = fragments.map(fragment => ({
    index: fragment.index,
    data: Array.from(fragment.data),
    metadata: {
      ...fragment.metadata,
      checksum: Array.from(fragment.metadata.checksum)
    }
  }));

  return JSON.stringify(serializable);
}

/**
 * Deserializes fragments from a portable format
 * @param serialized - Serialized fragments JSON string
 * @returns Deserialized fragments
 * @throws Error if deserialization fails
 */
export function deserializeFragments(serialized: string): Fragment[] {
  try {
    const data = JSON.parse(serialized);

    if (!Array.isArray(data)) {
      throw new Error('Invalid fragment data format');
    }

    return data.map(item => ({
      index: item.index,
      data: new Uint8Array(item.data),
      metadata: {
        ...item.metadata,
        checksum: new Uint8Array(item.metadata.checksum)
      }
    }));
  } catch (error) {
    throw new Error(`Failed to deserialize fragments: ${error}`);
  }
}

/**
 * Compresses fragments using simple run-length encoding
 * @param fragments - Fragments to compress
 * @returns Compressed fragments
 */
export function compressFragments(fragments: Fragment[]): Fragment[] {
  return fragments.map(fragment => ({
    ...fragment,
    data: runLengthEncode(fragment.data)
  }));
}

/**
 * Decompresses fragments using run-length decoding
 * @param fragments - Compressed fragments to decompress
 * @returns Decompressed fragments
 */
export function decompressFragments(fragments: Fragment[]): Fragment[] {
  return fragments.map(fragment => ({
    ...fragment,
    data: runLengthDecode(fragment.data)
  }));
}

// Helper functions for compression

function runLengthEncode(data: Uint8Array): Uint8Array {
  if (data.length === 0) return data;

  const encoded: number[] = [];
  let current = data[0]!;
  let count = 1;

  for (let i = 1; i < data.length; i++) {
    if (data[i] === current && count < 255) {
      count++;
    } else {
      encoded.push(count, current);
      current = data[i]!;
      count = 1;
    }
  }

  encoded.push(count, current);
  return new Uint8Array(encoded);
}

function runLengthDecode(encoded: Uint8Array): Uint8Array {
  if (encoded.length % 2 !== 0) {
    throw new Error('Invalid run-length encoded data');
  }

  const decoded: number[] = [];

  for (let i = 0; i < encoded.length; i += 2) {
    const count = encoded[i]!;
    const value = encoded[i + 1]!;

    for (let j = 0; j < count; j++) {
      decoded.push(value);
    }
  }

  return new Uint8Array(decoded);
}
